internal/credential/vault/vault_capabilities.go

```diff
@@ -0,0 +1,265 @@
+package vault
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type capabilities uint
+
+const noCapabilities = capabilities(0)
+
+const (
+	denyCapability capabilities = 1 << iota
+	createCapability
+	readCapability
+	updateCapability
+	deleteCapability
+	listCapability
+	sudoCapability
+	rootCapability
+)
+
+var string2Cap = map[string]capabilities{
+	"deny":   denyCapability,
+	"create": createCapability,
+	"read":   readCapability,
+	"update": updateCapability,
+	"delete": deleteCapability,
+	"list":   listCapability,
+	"sudo":   sudoCapability,
+	"root":   rootCapability,
+}
+
+func (c capabilities) goStrings() []string {
+	if c == noCapabilities {
+		return nil
+	}
+
+	var caps []string
+	if c&denyCapability == denyCapability {
+		caps = append(caps, "denyCapability")
+	}
+	if c&createCapability == createCapability {
+		caps = append(caps, "createCapability")
+	}
+	if c&readCapability == readCapability {
+		caps = append(caps, "readCapability")
+	}
+	if c&updateCapability == updateCapability {
+		caps = append(caps, "updateCapability")
+	}
+	if c&deleteCapability == deleteCapability {
+		caps = append(caps, "deleteCapability")
+	}
+	if c&listCapability == listCapability {
+		caps = append(caps, "listCapability")
+	}
+	if c&sudoCapability == sudoCapability {
+		caps = append(caps, "sudoCapability")
+	}
+	if c&rootCapability == rootCapability {
+		caps = append(caps, "rootCapability")
+	}
+	return caps
+}
+
+func (c capabilities) GoString() string {
+	if c == noCapabilities {
+		return "vault.capabilities{}"
+	}
+	caps := c.goStrings()
+	var b strings.Builder
+	b.WriteString("vault.capabilities{ ")
+	b.WriteString(strings.Join(caps, " | "))
+	b.WriteString(" }")
+	return b.String()
+}
+
+func (c capabilities) strings() []string {
+	if c == noCapabilities {
+		return nil
+	}
+
+	var caps []string
+	if c&denyCapability == denyCapability {
+		caps = append(caps, "deny")
+	}
+	if c&createCapability == createCapability {
+		caps = append(caps, "create")
+	}
+	if c&readCapability == readCapability {
+		caps = append(caps, "read")
+	}
+	if c&updateCapability == updateCapability {
+		caps = append(caps, "update")
+	}
+	if c&deleteCapability == deleteCapability {
+		caps = append(caps, "delete")
+	}
+	if c&listCapability == listCapability {
+		caps = append(caps, "list")
+	}
+	if c&sudoCapability == sudoCapability {
+		caps = append(caps, "sudo")
+	}
+	if c&rootCapability == rootCapability {
+		caps = append(caps, "root")
+	}
+
+	return caps
+}
+
+func (c capabilities) String() string {
```

Comment:
This is fine, but doing what you did in `GoString` and using strings.Join seems a bit more readable to me.  So something like:
```
	if c == noCapabilities {
		return "[]"
	}
	var b strings.Builder
	b.WriteString(`["`)
	b.WriteString(strings.Join(caps, `" ,"`))
	b.WriteString(`"]`)
	return b.String()
```

Reply:
Great idea. Ty!